clang/include/clang/Serialization/SourceLocationEncoding.h

```diff
@@ -149,14 +157,44 @@ class SourceLocationSequence::State {
   operator SourceLocationSequence *() { return &Seq; }
 };
 
-inline uint64_t SourceLocationEncoding::encode(SourceLocation Loc,
-                                               SourceLocationSequence *Seq) {
-  return Seq ? Seq->encode(Loc) : encodeRaw(Loc.getRawEncoding());
+inline SourceLocationEncoding::RawLocEncoding
+SourceLocationEncoding::encode(SourceLocation Loc, UIntTy BaseOffset,
+                               unsigned BaseModuleFileIndex,
+                               SourceLocationSequence *Seq) {
+  // If the source location is a local source location, we can try to optimize
+  // the similar sequences to only record the differences.
+  if (!BaseOffset)
+    return Seq ? Seq->encode(Loc) : encodeRaw(Loc.getRawEncoding());
+
+  if (Loc.isInvalid())
+    return 0;
+  
+  // Otherwise, the higher bits are used to store the module file index,
+  // so it is meaningless to optimize the source locations into small
+  // integers. Let's try to always use the raw encodings.
+  assert(Loc.getOffset() >= BaseOffset);
+  Loc = Loc.getLocWithOffset(-BaseOffset);
+  RawLocEncoding Encoded = encodeRaw(Loc.getRawEncoding());
+  assert(Encoded < ((RawLocEncoding)1 << 32));
```

Comment:
My compiler says:
```
warning: result of comparison of constant 4294967296 with expression of type 'unsigned int' is always true [-Wtautological-constant-out-of-range-compare]
```

Reply:
I guess you may refer to the next line and that one is indeed comparing unsigned with `(1<<32)`, and I fixed that. This line should be fine since Encoded has type `RawLocEncoding `.

Reply:
No, I'm referring to this line. I assume the compiler sees through `RawLocEncoding` and knows it's being initialized from `SourceLocation::UIntTy`.

Reply:
Oh, the compiler may be too smart : )

My intention was like, if someday we turned 64 bit source location on, this assertion can make us remember  to update here.

But given the warning, I don't know how to write the assertion here. So I just removed it.